src/charm.py

```diff
@@ -17,50 +17,164 @@
+    RESTART_MAP = {
+        '/usr/share/nvidia/vgpu/vgpuConfig.xml': ['nvidia-vgpu-mgr'],
+    }
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.framework.observe(self.on.install, self._on_install)
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        super().register_status_check(self.__check_status)
 
-    def _on_install(self, _):
-        """install hook."""
-        self.__set_ready_unit_status()
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.start, self._on_start)
 
-    def _on_upgrade_charm(self, _):
-        """upgrade-charm hook."""
-        self.__set_ready_unit_status()
+        # hash of the last successfully installed NVIDIA vGPU software passed
+        # as resource to the charm:
+        self._stored.set_default(last_installed_resource_hash=None)
 
     def _on_config_changed(self, _):
         """config-changed hook."""
-        if not self._has_nvidia_gpu_hardware():
-            return
+        # NOTE(lourot): We want to re-install the software here if a new
+        # version has just been provided as a charm resource.
+        self.__install_nvidia_software_if_needed()
 
         vgpu_device_mappings_str = self.config.get('vgpu-device-mappings')
         if vgpu_device_mappings_str is not None:
             vgpu_device_mappings = YAML().load(vgpu_device_mappings_str)
             logging.debug('vgpu-device-mappings={}'.format(
                 vgpu_device_mappings))
 
-    def __set_ready_unit_status(self):
-        """Set the unit status to active/ready."""
-        unit_status_msg = (
-            'Unit is ready: '
-            + ('no ' if not self._has_nvidia_gpu_hardware() else '')
-            + 'NVIDIA GPU found')
-        self.unit.status = ActiveStatus(unit_status_msg)
+        self.update_status()
+
+    def _on_start(self, _):
+        """start hook."""
+        # NOTE(lourot): We install software in the `start` hook instead of
+        # the `install` hook because we want to be able to install software
+        # after a reboot if NVIDIA hardware has then been added for the
+        # first time.
+        self.__install_nvidia_software_if_needed()
+
+        # NOTE(lourot): this is used by OSBaseCharm.update_status():
+        self._stored.is_started = True
+
+        self.update_status()
+
+    def services(self):
+        # If no NVIDIA software is expected to be installed on this particular
+        # unit, then no service should be expected to run by
+        # OSBaseCharm.update_status(). Otherwise the services from the
+        # RESTART_MAP are expected to run.
+        if not self.__is_nvidia_software_to_be_installed():
+            return []
+        return super().services()
+
+    def __check_status(self):
```

Comment:
Any particular reason for using dunder methods (double underscore). The name mangling is different which makes them impervious to overloading?

Reply:
I thought it was a good practice for exactly this reason, but now that you're making me reflect, it doesn't sound like a great idea after all. I'll change this, thanks!